File: backend/app/services/drift_position_fetcher.py
"""Fetch user positions from Drift Protocol and store snapshots.

Drift API endpoints used (base: https://data.api.drift.trade):
  - GET /authority/{wallet}/insuranceFundStake — IF stake events (last 31 days)
  - GET /authority/{wallet}/insuranceFundStake/{year}/{month} — historical IF events
  - GET /stats/insuranceFund — IF pool APYs
  - GET /stats/spotMarketAccounts — on-chain IF vault totals
  - GET /authority/{wallet}/snapshots/vaults?days={1,100} — strategy vault snapshots (multi-window)
"""
import logging
from collections import defaultdict
from datetime import datetime, timezone
from typing import Any, Optional

import httpx
from sqlalchemy.orm import Session

from app.models.base import SessionLocal
from app.models.user_position import TrackedWallet
from app.services.utils import (
    safe_float, get_or_none, cached, parse_timestamp, compute_realized_apy,
    load_opportunity_map, compute_held_days, build_position_dict,
    store_events_batch,
)

logger = logging.getLogger(__name__)

DRIFT_API = "https://data.api.drift.trade"


def _get(path: str, client: httpx.Client) -> Optional[dict | list]:
    return get_or_none(f"{DRIFT_API}{path}", client, log_label="Drift API")


_float = safe_float
_cached = cached
_parse_timestamp = parse_timestamp
_held_days = compute_held_days
_pos = build_position_dict


# ---------------------------------------------------------------------------
# IF pool state (shared across all wallets)
# ---------------------------------------------------------------------------

def _get_if_pool_apys(client: httpx.Client) -> dict[int, float]:
    """Fetch IF pool APYs: {marketIndex: apy_pct}."""
    def _fetch():
        data = _get("/stats/insuranceFund", client)
        if not isinstance(data, dict):
            return {}
        result = {}
        for entry in data.get("marketSharePriceData", []):
            idx = entry.get("marketIndex")
            apy = _float(entry.get("apy"))
            if idx is not None and apy is not None:
                result[idx] = apy
        return result
    return _cached("drift_if_apys", 300, _fetch)


# ---------------------------------------------------------------------------
# Insurance Fund positions
# ---------------------------------------------------------------------------

def _fetch_if_events(
    wallet: str, client: httpx.Client,
) -> list[dict]:
    """Fetch IF stake events. Tries recent 31 days first, then goes back month by month."""
    raw = _get(f"/authority/{wallet}/insuranceFundStake", client)
    events = raw.get("records", []) if isinstance(raw, dict) else (raw if isinstance(raw, list) else [])
    if events:
        return events

    # Search backwards month by month (up to 6 months)
    now = datetime.now(timezone.utc)
    all_events = []
    for months_back in range(0, 7):
        year = now.year
        month = now.month - months_back
        while month <= 0:
            month += 12
            year -= 1
        raw = _get(f"/authority/{wallet}/insuranceFundStake/{year}/{month}", client)
        monthly = raw.get("records", []) if isinstance(raw, dict) else (raw if isinstance(raw, list) else [])
        if monthly:
            all_events.extend(monthly)

    return all_events


def _fetch_if_positions(
    wallet: str, client: httpx.Client, db: Session, now: datetime,
    opp_map: dict | None = None,
) -> tuple[list[dict], list[dict]]:
    """Fetch Insurance Fund staking positions. Returns (positions, events)."""
    events = _fetch_if_events(wallet, client)
    if not events:
        return [], []

    if_apys = _get_if_pool_apys(client)

    # Group events by marketIndex, keep latest per market
    by_market: dict[int, list[dict]] = defaultdict(list)
    for evt in events:
        idx = evt.get("marketIndex")
        if idx is not None:
            by_market[idx].append(evt)

    positions = []
    position_events = []

    for market_index, market_events in by_market.items():
        market_events.sort(key=lambda e: e.get("ts", 0))
        latest = market_events[-1]

        shares_after = _float(latest.get("ifSharesAfter")) or 0.0
        if shares_after < 0.001:
            # Position is closed — still record events but skip position
            for evt in market_events:
                position_events.append(_if_event_to_record(evt, wallet))
            continue

        symbol = latest.get("symbol", f"MARKET-{market_index}")
        external_id = f"drift-if-{market_index}"

        # Current value: proportional share of IF vault (stablecoins — token units ≈ USD).
        total_shares_ev = _float(latest.get("totalIfSharesAfter"))
        vault_amount_ev = _float(latest.get("insuranceVaultAmountBefore"))
        deposit_amount_usd = (
            (shares_after / total_shares_ev) * vault_amount_ev
            if total_shares_ev and total_shares_ev > 0 and vault_amount_ev
            else None
        )

        total_staked = 0.0
        total_unstaked = 0.0
        opened_at = None
        for evt in market_events:
            action = (evt.get("action") or "").lower()
            amount = _float(evt.get("amount")) or 0.0
            if action == "stake":
                total_staked += amount
                if opened_at is None:
                    opened_at = _parse_timestamp(evt.get("ts"))
            elif action in ("unstake", "unstakerequest"):
                total_unstaked += amount

        net_staked = total_staked - total_unstaked
        initial_deposit_usd = net_staked if net_staked > 0 else total_staked

        pnl_usd = (deposit_amount_usd - initial_deposit_usd) if deposit_amount_usd is not None and initial_deposit_usd > 0 else None
        pnl_pct = (pnl_usd / initial_deposit_usd * 100) if pnl_usd is not None else None
        held_days = _held_days(opened_at, now)

        apy = if_apys.get(market_index)

        _omap = opp_map if opp_map is not None else load_opportunity_map(db)
        entry = _omap.get(external_id)
        opportunity_id = entry["id"] if entry else None
        if apy is None and entry:
            apy = entry["apy_current"]

        positions.append(_pos(
            wallet_address=wallet, protocol_slug="drift",
            product_type="insurance_fund", external_id=external_id,
            snapshot_at=now, opportunity_id=opportunity_id,
            deposit_amount=shares_after, deposit_amount_usd=deposit_amount_usd,
            pnl_usd=pnl_usd, pnl_pct=pnl_pct,
            initial_deposit_usd=initial_deposit_usd,
            opened_at=opened_at, held_days=held_days, apy=apy,
            token_symbol=symbol,
            extra_data={
                "if_shares": shares_after, "market_index": market_index,
                "symbol": symbol, "total_staked": total_staked,
                "total_unstaked": total_unstaked,
            },
        ))

        for evt in market_events:
            position_events.append(_if_event_to_record(evt, wallet))

    return positions, position_events


def _if_event_to_record(evt: dict, wallet: str) -> dict:
    """Convert a Drift IF stake event to UserPositionEvent format."""
    action = (evt.get("action") or "unknown").lower()
    market_index = evt.get("marketIndex", 0)
    return {
        "wallet_address": wallet,
        "protocol_slug": "drift",
        "product_type": "insurance_fund",
        "external_id": f"drift-if-{market_index}",
        "event_type": action,
        "amount": _float(evt.get("amount")),
        "amount_usd": _float(evt.get("amount")),  # IF stakes are in the token itself
        "tx_signature": evt.get("txSig"),
        "event_at": _parse_timestamp(evt.get("ts")) or datetime.now(timezone.utc),
        "extra_data": {
            "symbol": evt.get("symbol"),
            "market_index": market_index,
            "if_shares_before": _float(evt.get("ifSharesBefore")),
            "if_shares_after": _float(evt.get("ifSharesAfter")),
        },
    }


# ---------------------------------------------------------------------------
# Strategy Vault positions
# ---------------------------------------------------------------------------

def _fetch_vault_positions(
    wallet: str, client: httpx.Client, db: Session, now: datetime,
    opp_map: dict | None = None,
) -> list[dict]:
    """Fetch Strategy Vault positions from daily snapshots.

    Fetches from multiple time windows (1 day + 100 days) because the Drift API
    partitions data differently by time range — recent vaults may only appear in
    short windows. Deduplicates and filters stale positions.
    """
    STALE_DAYS = 7

    # Fetch from multiple windows to catch both recent and historical vaults
    by_vault: dict[str, list[dict]] = defaultdict(list)
    seen_keys: set[tuple[str, Any]] = set()  # (vault, ts) for dedup

    for days in (1, 100):
        raw = _get(f"/authority/{wallet}/snapshots/vaults?days={days}", client)
        if not isinstance(raw, dict):
            continue
        for account in raw.get("accounts", []):
            for snap in account.get("snapshots", []):
                vault = snap.get("vault")
                if not vault:
                    continue
                dedup_key = (vault, snap.get("ts"))
                if dedup_key in seen_keys:
                    continue
                seen_keys.add(dedup_key)
                by_vault[vault].append(snap)

    if not by_vault:
        return []

    positions = []
    for vault_pubkey, snapshots in by_vault.items():
        snapshots.sort(key=lambda s: s.get("ts", 0))
        latest = snapshots[-1]

        # Skip stale positions — user likely withdrew
        latest_ts = _parse_timestamp(latest.get("ts"))
        if latest_ts and (now - latest_ts).total_seconds() > STALE_DAYS * 86400:
            logger.info("Skipping stale vault %s (last snapshot %s)", vault_pubkey[:12], latest_ts)
            continue

        total_value = _float(latest.get("totalAccountValue")) or 0.0
        if total_value <= 0:
            continue

        net_deposits = _float(latest.get("netDeposits")) or 0.0
        market_index = latest.get("marketIndex")

        pnl_usd = total_value - net_deposits
        pnl_pct = (pnl_usd / net_deposits * 100) if net_deposits > 0 else None

        external_id = f"drift-vault-{vault_pubkey}"

        _omap = opp_map if opp_map is not None else load_opportunity_map(db)
        entry = _omap.get(vault_pubkey) or _omap.get(external_id)
        opportunity_id = entry["id"] if entry else None
        apy = entry["apy_current"] if entry else None
        token_symbol = entry["first_token"] if entry else None

        if not token_symbol:
            token_symbol = f"MARKET-{market_index}" if market_index is not None else None

        # Estimate opened_at from earliest snapshot
        opened_at = _parse_timestamp(snapshots[0].get("ts"))
        held_days = _held_days(opened_at, now)

        positions.append(_pos(
            wallet_address=wallet, protocol_slug="drift",
            product_type="earn_vault", external_id=external_id,
            snapshot_at=now, opportunity_id=opportunity_id,
            deposit_amount=total_value, deposit_amount_usd=total_value,
            pnl_usd=pnl_usd, pnl_pct=pnl_pct,
            initial_deposit_usd=net_deposits if net_deposits > 0 else None,
            opened_at=opened_at, held_days=held_days, apy=apy,
            token_symbol=token_symbol,
            extra_data={
                "vault_pubkey": vault_pubkey, "market_index": market_index,
                "net_deposits": net_deposits,
                "total_account_value": total_value,
                "total_account_base_value": _float(latest.get("totalAccountBaseValue")),
            },
        ))

    return positions


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------

def fetch_wallet_positions(wallet_address: str, db: Session) -> dict:
    """Fetch all Drift positions for a wallet. Returns dict with positions list."""
    now = datetime.now(timezone.utc)
    opp_map = load_opportunity_map(db)

    with httpx.Client() as client:
        if_positions, if_events = _fetch_if_positions(wallet_address, client, db, now, opp_map=opp_map)
        vault_positions = _fetch_vault_positions(wallet_address, client, db, now, opp_map=opp_map)

    all_positions = if_positions + vault_positions

    total_value_usd = sum(_float(p.get("deposit_amount_usd")) or 0 for p in all_positions)
    total_pnl_usd = sum(_float(p.get("pnl_usd")) or 0 for p in all_positions if p.get("pnl_usd") is not None)

    return {
        "wallet": wallet_address,
        "positions": all_positions,
        "events": if_events,
        "summary": {
            "total_value_usd": total_value_usd,
            "total_pnl_usd": total_pnl_usd,
            "position_count": len(all_positions),
        },
    }


def fetch_wallet_events(wallet_address: str) -> list[dict]:
    """Fetch IF stake/unstake events for a wallet."""
    with httpx.Client() as client:
        events = _fetch_if_events(wallet_address, client)

    return [_if_event_to_record(evt, wallet_address) for evt in events]


# ---------------------------------------------------------------------------
# Background job: snapshot all tracked wallets
# ---------------------------------------------------------------------------

def snapshot_all_wallets(db: Session, snapshot_at: datetime | None = None) -> int:
    """Iterate all active TrackedWallets, fetch Drift positions, store snapshots."""
    wallets = (
        db.query(TrackedWallet)
        .filter(TrackedWallet.is_active.is_(True))
        .all()
    )
    if not wallets:
        logger.info("No tracked wallets for Drift snapshot")
        return 0

    logger.info("Snapshotting Drift positions for %d wallets", len(wallets))
    now = snapshot_at or datetime.now(timezone.utc)
    total_snapshots = 0
    opp_map = load_opportunity_map(db)

    with httpx.Client() as client:
        _get_if_pool_apys(client)

        for wallet in wallets:
            try:
                if_positions, if_events = _fetch_if_positions(
                    wallet.wallet_address, client, db, now, opp_map=opp_map,
                )
                vault_positions = _fetch_vault_positions(
                    wallet.wallet_address, client, db, now, opp_map=opp_map,
                )
                all_positions = if_positions + vault_positions

                from app.services.utils import store_position_rows
                total_snapshots += store_position_rows(db, all_positions, now)

                store_events_batch(db, if_events)

                wallet.last_fetched_at = now
                db.flush()

                logger.info(
                    "Drift wallet %s: %d positions snapshotted",
                    wallet.wallet_address[:8],
                    len(all_positions),
                )

            except Exception as exc:
                logger.error(
                    "Failed to snapshot Drift wallet %s: %s",
                    wallet.wallet_address[:8],
                    exc,
                )
                continue

    db.commit()
    logger.info("Drift position snapshot complete: %d total snapshots", total_snapshots)
    return total_snapshots

File: backend/src/monitor/services/drift-position-fetcher.ts
/**
 * Fetch user positions from Drift Protocol and store snapshots.
 *
 * Port of backend/app/services/drift_position_fetcher.py
 *
 * Two position types:
 *   - Insurance Fund staking: IF stake events → proportional share
 *   - Strategy Vaults: daily snapshots → totalAccountValue vs netDeposits
 */
import type { NodePgDatabase } from "drizzle-orm/node-postgres";
import { eq } from "drizzle-orm";
import { getOrNull } from "../../shared/http.js";
import { logger } from "../../shared/logger.js";
import { safeFloat, parseTimestamp, cached } from "../../shared/utils.js";
import type { OpportunityMapEntry } from "../../shared/types.js";
import { discoverService } from "../../discover/service.js";
import { db } from "../db/connection.js";
import { trackedWallets } from "../db/schema.js";
import {
  buildPositionDict,
  computeHeldDays,
  storePositionRows,
  storeEventsBatch,
  loadOpportunityMap,
  type PositionDict,
  type EventDict,
} from "./utils.js";

const DRIFT_API = "https://data.api.drift.trade";

async function dGet(path: string): Promise<unknown | null> {
  return getOrNull(`${DRIFT_API}${path}`, { logLabel: "Drift API" });
}

// ---------------------------------------------------------------------------
// IF pool APYs (cached 5 min)
// The /stats/insuranceFund API only returns { marketIndex, symbol, apy }
// per entry — no share price or vault balance fields.
// ---------------------------------------------------------------------------

let _ifApyCache: { at: number; data: Record<number, number> } | null = null;

async function getIfPoolApys(): Promise<Record<number, number>> {
  const now = Date.now();
  if (_ifApyCache && now - _ifApyCache.at < 300_000) return _ifApyCache.data;

  const data = await dGet("/stats/insuranceFund");
  const result: Record<number, number> = {};
  if (data && typeof data === "object" && !Array.isArray(data)) {
    const entries =
      ((data as Record<string, unknown>).data as Record<string, unknown>)
        ?.marketSharePriceData ??
      (data as Record<string, unknown>).marketSharePriceData;
    if (Array.isArray(entries)) {
      for (const e of entries as Record<string, unknown>[]) {
        const idx = e.marketIndex;
        const apy = safeFloat(e.apy);
        if (idx !== undefined && apy !== null) result[Number(idx)] = apy;
      }
    }
  }
  _ifApyCache = { at: now, data: result };
  return result;
}

// ---------------------------------------------------------------------------
// IF events
// ---------------------------------------------------------------------------

async function fetchIfEvents(wallet: string): Promise<Record<string, unknown>[]> {
  const raw = await dGet(`/authority/${wallet}/insuranceFundStake`);
  let events: Record<string, unknown>[] = [];

  if (raw && typeof raw === "object" && !Array.isArray(raw)) {
    events = ((raw as Record<string, unknown>).records as Record<string, unknown>[]) ?? [];
  } else if (Array.isArray(raw)) {
    events = raw as Record<string, unknown>[];
  }

  if (events.length > 0) return events;

  // Fallback: search month by month (up to 6 months back)
  const now = new Date();
  const allEvents: Record<string, unknown>[] = [];
  for (let monthsBack = 0; monthsBack < 7; monthsBack++) {
    let year = now.getFullYear();
    let month = now.getMonth() + 1 - monthsBack;
    while (month <= 0) {
      month += 12;
      year--;
    }
    const raw2 = await dGet(
      `/authority/${wallet}/insuranceFundStake/${year}/${month}`,
    );
    let monthly: Record<string, unknown>[] = [];
    if (raw2 && typeof raw2 === "object" && !Array.isArray(raw2)) {
      monthly = ((raw2 as Record<string, unknown>).records as Record<string, unknown>[]) ?? [];
    } else if (Array.isArray(raw2)) {
      monthly = raw2 as Record<string, unknown>[];
    }
    if (monthly.length > 0) allEvents.push(...monthly);
  }
  return allEvents;
}

// ---------------------------------------------------------------------------
// IF positions
// ---------------------------------------------------------------------------

function ifEventToRecord(evt: Record<string, unknown>, wallet: string): EventDict {
  const action = ((evt.action as string) ?? "unknown").toLowerCase();
  const marketIndex = evt.marketIndex ?? 0;
  return {
    wallet_address: wallet,
    protocol_slug: "drift",
    product_type: "insurance_fund",
    external_id: `drift-if-${marketIndex}`,
    event_type: action,
    amount: safeFloat(evt.amount),
    amount_usd: safeFloat(evt.amount),
    tx_signature: (evt.txSig as string) ?? null,
    event_at: parseTimestamp(evt.ts) ?? new Date(),
    extra_data: {
      symbol: evt.symbol,
      market_index: marketIndex,
      if_shares_before: safeFloat(evt.ifSharesBefore),
      if_shares_after: safeFloat(evt.ifSharesAfter),
    },
  };
}

async function fetchIfPositions(
  wallet: string,
  now: Date,
  oppMap: Record<string, OpportunityMapEntry>,
): Promise<{ positions: PositionDict[]; events: EventDict[] }> {
  const events = await fetchIfEvents(wallet);
  if (events.length === 0) return { positions: [], events: [] };

  const ifApys = await getIfPoolApys();

  // Group by marketIndex
  const byMarket: Record<number, Record<string, unknown>[]> = {};
  for (const evt of events) {
    const idx = Number(evt.marketIndex ?? -1);
    if (idx < 0) continue;
    if (!byMarket[idx]) byMarket[idx] = [];
    byMarket[idx].push(evt);
  }

  const positions: PositionDict[] = [];
  const positionEvents: EventDict[] = [];

  for (const [marketIndex, marketEvents] of Object.entries(byMarket)) {
    const idx = Number(marketIndex);
    marketEvents.sort((a, b) => Number(a.ts ?? 0) - Number(b.ts ?? 0));
    const latest = marketEvents[marketEvents.length - 1];

    const sharesAfter = safeFloat(latest.ifSharesAfter) ?? 0;
    if (sharesAfter < 0.001) {
      // Closed — record events only
      for (const evt of marketEvents)
        positionEvents.push(ifEventToRecord(evt, wallet));
      continue;
    }

    const symbol =
      (latest.symbol as string) ?? `MARKET-${idx}`;
    const externalId = `drift-if-${idx}`;
    const entry = oppMap[externalId];

    // Compute current deposit value using opportunity TVL (updated every
    // 15 min from on-chain data) and event-level total shares.
    // The Drift /stats/insuranceFund API only returns { marketIndex, symbol, apy }
    // — no share price or vault balance — so we derive value from the Discover
    // module's TVL combined with the most recent event's total shares.
    const totalSharesEv = safeFloat(latest.totalIfSharesAfter);
    const tvlUsd = entry?.tvl_usd ?? null;
    let depositAmountUsd: number | null = null;

    if (tvlUsd && tvlUsd > 0 && totalSharesEv && totalSharesEv > 0) {
      // TVL is current (from on-chain); totalShares from last event is approximate
      depositAmountUsd = (sharesAfter / totalSharesEv) * tvlUsd;
    }

    // Fallback: event-level vault amount (raw token units → USD)
    if (depositAmountUsd === null) {
      const vaultAmountRaw = safeFloat(latest.insuranceVaultAmountBefore);
      if (totalSharesEv && totalSharesEv > 0 && vaultAmountRaw) {
        // insuranceVaultAmountBefore is in raw token units (6 decimals for stablecoins)
        const STABLECOIN_DECIMALS = 6;
        const vaultAmountUsd = vaultAmountRaw / 10 ** STABLECOIN_DECIMALS;
        depositAmountUsd = (sharesAfter / totalSharesEv) * vaultAmountUsd;
        logger.debug(
          { wallet: wallet.slice(0, 8), market: idx, vaultAmountRaw, vaultAmountUsd, deposit: depositAmountUsd },
          "Drift IF: using normalized event fallback for deposit value",
        );
      }
    }

    if (depositAmountUsd === null) {
      logger.info(
        { wallet: wallet.slice(0, 8), market: idx, shares: sharesAfter },
        "Drift IF: could not compute current value — no TVL or event data",
      );
    }

    let totalStaked = 0;
    let totalUnstaked = 0;
    let openedAt: Date | null = null;

    for (const evt of marketEvents) {
      const action = ((evt.action as string) ?? "").toLowerCase();
      const amount = safeFloat(evt.amount) ?? 0;
      if (action === "stake") {
        totalStaked += amount;
        if (openedAt === null) openedAt = parseTimestamp(evt.ts);
      } else if (action === "unstake" || action === "unstakerequest") {
        totalUnstaked += amount;
      }
    }

    const netStaked = totalStaked - totalUnstaked;
    const initialDepositUsd = netStaked > 0 ? netStaked : totalStaked;
    const pnlUsd =
      depositAmountUsd !== null && initialDepositUsd > 0
        ? depositAmountUsd - initialDepositUsd
        : null;
    const pnlPct =
      pnlUsd !== null && initialDepositUsd > 0
        ? (pnlUsd / initialDepositUsd) * 100
        : null;

    let apy: number | null = ifApys[idx] ?? null;
    if (apy === null && entry) apy = entry.apy_current;

    positions.push(
      buildPositionDict({
        wallet_address: wallet,
        protocol_slug: "drift",
        product_type: "insurance_fund",
        external_id: externalId,
        snapshot_at: now,
        opportunity_id: entry?.id ?? null,
        deposit_amount: sharesAfter,
        deposit_amount_usd: depositAmountUsd,
        pnl_usd: pnlUsd,
        pnl_pct: pnlPct,
        initial_deposit_usd: initialDepositUsd > 0 ? initialDepositUsd : null,
        opened_at: openedAt,
        held_days: computeHeldDays(openedAt, now),
        apy,
        token_symbol: symbol,
        extra_data: {
          if_shares: sharesAfter,
          market_index: idx,
          symbol,
          total_staked: totalStaked,
          total_unstaked: totalUnstaked,
        },
      }),
    );

    for (const evt of marketEvents)
      positionEvents.push(ifEventToRecord(evt, wallet));
  }

  return { positions, events: positionEvents };
}

// ---------------------------------------------------------------------------
// Strategy vault positions
// ---------------------------------------------------------------------------

async function fetchVaultPositions(
  wallet: string,
  now: Date,
  oppMap: Record<string, OpportunityMapEntry>,
): Promise<PositionDict[]> {
  const STALE_DAYS = 7;
  const byVault: Record<string, Record<string, unknown>[]> = {};
  const seenKeys = new Set<string>();

  for (const days of [1, 100]) {
    const raw = await dGet(
      `/authority/${wallet}/snapshots/vaults?days=${days}`,
    );
    if (!raw || typeof raw !== "object" || Array.isArray(raw)) continue;

    const accounts =
      ((raw as Record<string, unknown>).accounts as Record<
        string,
        unknown
      >[]) ?? [];
    for (const account of accounts) {
      const snapshots =
        (account.snapshots as Record<string, unknown>[]) ?? [];
      for (const snap of snapshots) {
        const vault = snap.vault as string;
        if (!vault) continue;
        const key = `${vault}:${snap.ts}`;
        if (seenKeys.has(key)) continue;
        seenKeys.add(key);
        if (!byVault[vault]) byVault[vault] = [];
        byVault[vault].push(snap);
      }
    }
  }

  if (Object.keys(byVault).length === 0) return [];

  const positions: PositionDict[] = [];

  for (const [vaultPubkey, snapshots] of Object.entries(byVault)) {
    snapshots.sort((a, b) => Number(a.ts ?? 0) - Number(b.ts ?? 0));
    const latest = snapshots[snapshots.length - 1];

    const latestTs = parseTimestamp(latest.ts);
    if (
      latestTs &&
      (now.getTime() - latestTs.getTime()) / 1000 > STALE_DAYS * 86400
    ) {
      continue;
    }

    const totalValue = safeFloat(latest.totalAccountValue) ?? 0;
    if (totalValue <= 0) continue;

    const netDeposits = safeFloat(latest.netDeposits) ?? 0;
    const marketIndex = latest.marketIndex;
    const pnlUsd = totalValue - netDeposits;
    const pnlPct = netDeposits > 0 ? (pnlUsd / netDeposits) * 100 : null;

    logger.debug(
      { wallet: wallet.slice(0, 8), vault: vaultPubkey.slice(0, 8), totalValue, netDeposits, pnlUsd, snapshotCount: snapshots.length },
      "Drift vault position computed",
    );

    const externalId = `drift-vault-${vaultPubkey}`;
    const entry =
      oppMap[vaultPubkey] ?? oppMap[externalId] ?? null;
    let tokenSymbol = entry?.first_token ?? null;
    if (!tokenSymbol && marketIndex !== undefined) {
      tokenSymbol = `MARKET-${marketIndex}`;
    }

    const openedAt = parseTimestamp(snapshots[0].ts);

    positions.push(
      buildPositionDict({
        wallet_address: wallet,
        protocol_slug: "drift",
        product_type: "earn_vault",
        external_id: externalId,
        snapshot_at: now,
        opportunity_id: entry?.id ?? null,
        deposit_amount: totalValue,
        deposit_amount_usd: totalValue,
        pnl_usd: pnlUsd,
        pnl_pct: pnlPct,
        initial_deposit_usd: netDeposits > 0 ? netDeposits : null,
        opened_at: openedAt,
        held_days: computeHeldDays(openedAt, now),
        apy: entry?.apy_current ?? null,
        token_symbol: tokenSymbol,
        extra_data: {
          vault_pubkey: vaultPubkey,
          market_index: marketIndex,
          net_deposits: netDeposits,
          total_account_value: totalValue,
          total_account_base_value: safeFloat(latest.totalAccountBaseValue),
        },
      }),
    );
  }

  return positions;
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

export async function fetchWalletPositions(
  walletAddress: string,
): Promise<{
  wallet: string;
  positions: PositionDict[];
  events: EventDict[];
  summary: { total_value_usd: number; total_pnl_usd: number; position_count: number };
}> {
  const now = new Date();
  const oppMap = await loadOpportunityMap(discoverService);

  const { positions: ifPositions, events: ifEvents } =
    await fetchIfPositions(walletAddress, now, oppMap);
  const vaultPositions = await fetchVaultPositions(
    walletAddress,
    now,
    oppMap,
  );

  const allPositions = [...ifPositions, ...vaultPositions];
  const totalValue = allPositions.reduce(
    (s, p) => s + (p.deposit_amount_usd ?? 0),
    0,
  );
  const totalPnl = allPositions.reduce(
    (s, p) => s + (p.pnl_usd ?? 0),
    0,
  );

  return {
    wallet: walletAddress,
    positions: allPositions,
    events: ifEvents,
    summary: {
      total_value_usd: totalValue,
      total_pnl_usd: totalPnl,
      position_count: allPositions.length,
    },
  };
}

export async function snapshotAllWallets(
  database: NodePgDatabase,
  snapshotAt?: Date,
): Promise<number> {
  const wallets = await database
    .select()
    .from(trackedWallets)
    .where(eq(trackedWallets.is_active, true));

  if (wallets.length === 0) return 0;

  logger.info({ count: wallets.length }, "Snapshotting Drift positions");
  const now = snapshotAt ?? new Date();
  let totalSnapshots = 0;
  const oppMap = await loadOpportunityMap(discoverService);
  await getIfPoolApys(); // prime cache

  for (const wallet of wallets) {
    try {
      const { positions: ifPositions, events: ifEvents } =
        await fetchIfPositions(wallet.wallet_address, now, oppMap);
      const vaultPositions = await fetchVaultPositions(
        wallet.wallet_address,
        now,
        oppMap,
      );

      const allPositions = [...ifPositions, ...vaultPositions];
      totalSnapshots += await storePositionRows(database, allPositions, now);
      await storeEventsBatch(database, ifEvents);

      await database
        .update(trackedWallets)
        .set({ last_fetched_at: now })
        .where(eq(trackedWallets.id, wallet.id));

      logger.info(
        {
          wallet: wallet.wallet_address.slice(0, 8),
          count: allPositions.length,
        },
        "Drift wallet snapshotted",
      );
    } catch (err) {
      logger.error(
        { err, wallet: wallet.wallet_address.slice(0, 8) },
        "Failed to snapshot Drift wallet",
      );
    }
  }

  logger.info({ totalSnapshots }, "Drift position snapshot complete");
  return totalSnapshots;
}
